crates/holochain_persistence_lmdb/src/cas/lmdb.rs

```diff
@@ -33,55 +31,109 @@ impl Debug for LmdbStorage {
+) -> Result<Option<Content>, StoreError> {
+    match result {
+        Ok(Some(Value::Json(s))) => Ok(Some(serde_json::from_str(&s).unwrap())),
+        Ok(None) => Ok(None),
+        Ok(Some(_v)) => Err(StoreError::DataError(rkv::DataError::UnexpectedType {
+            actual: rkv::value::Type::Json,
+            expected: rkv::value::Type::Json,
+        })),
+        Err(e) => Err(e),
+    }
+}
+
+fn handle_cursor_tuple_result(
+    result: Result<(&[u8], Option<rkv::Value>), StoreError>,
+) -> Result<(Address, Option<Content>), StoreError> {
+    match result {
+        Ok((address, Some(Value::Json(s)))) => Ok((
+            HashString::from(address.to_vec()),
+            Some(serde_json::from_str(&s).unwrap()),
+        )),
+        Ok((address, None)) => Ok((HashString::from(address.to_vec()), None)),
+        Ok((_address, Some(_v))) => Err(StoreError::DataError(rkv::DataError::UnexpectedType {
+            actual: rkv::value::Type::Json,
+            expected: rkv::value::Type::Json,
+        })),
+        Err(e) => Err(e),
+    }
+}
+
 impl LmdbStorage {
-    fn lmdb_add(&mut self, content: &dyn AddressableContent) -> Result<(), StoreError> {
+    pub fn lmdb_add<'env>(
+        &self,
+        mut writer: &mut rkv::Writer<'env>,
+        content: &dyn AddressableContent,
+    ) -> Result<(), StoreError> {
         self.lmdb.add(
+            &mut writer,
             content.address(),
             &Value::Json(&content.content().to_string()),
         )
     }
 
-    fn lmdb_fetch(&self, address: &Address) -> Result<Option<Content>, StoreError> {
-        let env = self.lmdb.manager.read().unwrap();
-        let reader = env.read()?;
+    pub fn lmdb_fetch(
+        &self,
+        reader: &Reader,
+        address: &Address,
+    ) -> Result<Option<Content>, StoreError> {
+        let result = self.lmdb.store.get(reader, address.clone());
+        handle_cursor_result(result)
+    }
 
-        match self.lmdb.store.get(&reader, address.clone()) {
-            Ok(Some(value)) => match value {
-                Value::Json(s) => Ok(Some(JsonString::from_json(s))),
-                _ => Err(StoreError::DataError(DataError::Empty)),
-            },
-            Ok(None) => Ok(None),
-            Err(e) => Err(e),
-        }
+    pub fn lmdb_iter(
+        &self,
+        reader: &Reader,
+    ) -> Result<Vec<(Address, Option<Content>)>, StoreError> {
+        self.lmdb
+            .store
+            .iter_start(reader)?
+            .map(handle_cursor_tuple_result)
+            .collect::<Result<Vec<(Address, Option<Content>)>, StoreError>>()
     }
 }
 
 impl ContentAddressableStorage for LmdbStorage {
-    fn add(&mut self, content: &dyn AddressableContent) -> PersistenceResult<()> {
-        self.lmdb_add(content)
+    fn add(&self, content: &dyn AddressableContent) -> PersistenceResult<()> {
+        let rkv = self.lmdb.rkv.write().unwrap();
+        let mut writer: Writer = rkv.write().unwrap(); // TODO Use ?
```

Comment:
yes to `?`!

Reply:
in general I agree. Is failing to acquire a write-lock on the `Rkv` environment a recoverable error though? It seems to me like that's either a bug or caused by outside irreversable tampering. If I'm right that this error is unrecoverable, then I think `unwrap()` or even better `expect(...)` is correct.